### ISSUE TITLE
Query Parameter Support

### ISSUE THREAD

**Issue by BernhardLenz** [restsql/restsql#72]
Does RestSql support or is there any plan to support query parameters in a where clause for select statements? 
e.g. 
select * from account where acct_num = ?

**Comment by restsql** [comment]
Absolutely. The tool's main purpose is to full CRUD including simple query parameterization. The main resource query is not parameterized, rather this is done via the http query params (and optionally path for a Pk query). Have you looked at the API reference: http://restsql.org/doc/ref/index.html. Look for the two GET operations.

**Comment by BernhardLenz** [comment]
Hi there. Yes, I looked at the GET operations but wasn't sure if I
understood the documentation correctly.

The way I interpreted the documentation is that the {name}={value}
parameters are appended to the where clause.

What I'm looking for is a parameter which can be embedded inside the sql
statement similar to a prepared statements ? notation. Some of the queries
are rather complicated (e.g. using subqueries in the from or where clause)
and therefore appending doesn't work in some cases.

A simplified example is
select *
from a
where a.id in (select b.id from b where b.val = ?)

Is this something that can be done with restsql?

**Comment by restsql** [comment]
Nope. My currently. You must hardcode any from/where clause values in the base query.

Just a thought - are you aware on the IN operator support? You can send in a list of values inside parens.

You are welcome to contribute framework extensions to support those subquery scenarios. I'm sure others will benefit.

Mark

**Comment by BernhardLenz** [comment]
Hi Mark. 

the result of the sub query can potentially be very large which could exceed the length restrictions of a sql query if used in an in statement. 

I'll have a look at it next week to see how easy it is to extend the framework. You guys use pull requests for that?

Cheers
Bernie

**Comment by restsql** [comment]
Understood. It's not full substitute for subselects.

Yes I can take in a fork easily through github, which is full git.

The framework developers guide in the sdk has some tips on Dev env and guidelines to follow. https://github.com/restsql/restsql/blob/master/README-FRAMEWORK_DEVELOPERS.txt.

This won't be as simple. Some concerns come to mind.

The resource metadata is lazy loaded from the definition file once. The main quiet is executed with a tacked on limit clause and the column metadata is inspected. If the subselect is in the main query, it needs the param value to make it work. You could put that 'default' value as a new element in the resource definition. (It already uses prepared statements exclusively, for security's sake). The main query has to actually return something. Don't think it will work on an empty query.

Alternatively the subselect could be injected at steady state. The main query could be there to describe the columns with a simple query. Some kind of marker could be used to indicate the sub select to be dropped into the statement.

Also the parameters for a request a have to be distinguished between the normal select columns and the sub select columns. Probably have to call this out with some new element in the definition. It will have to be uniquely named or else the request post params/URL parsing will be nondeterminate. Or some kind of special char in the param name could be used. Like $val=xxx. That assumes all those are unique. So you could use $1=xxx. Or $SUB1=xxx.

Mark

**Comment by BernhardLenz** [comment]
Hi Mark.

What I was thinking is to maybe extend the "Special parameters:
_limit={integer},
_offset={integer}, _output={media-type}" by a parameter named _q1={value},
_q2={value}, _q3={value}, etc which corresponds to the first, second, third
question mark in the prepared statement. restsql would then iterate over
the _q parameters and call the setObject(int parameterIndex, Object x)
method of the prepared statement.

Is this approach in line with the architecture of restsql?

Bernie

**Comment by restsql** [comment]
Your naming suggestion is aligned with existing special params. q# is okay but not very descriptive. Does q mean query? How about _param1, _param2, etc?

We have to set the value in at least two cases - 1) initial main query that is automatic and 2) a steady state
query. So the initial one needs a default value, right? What if the initial request is a write? Would writes need the params? I don't have the vision to see all the possibilities of the main query.

The AbstractSqlBuilder is where the setObject on the query would happen but we would need to also extend the AbstractSqlResourceMetaData to pull in metadata for the default value. And I think RequestValue to recognize the special request param.

**Comment by BernhardLenz** [comment]
I do like _parm# !!!

Hmmm. the initial query throws a curve ball. Defaulting could be tough,
since we don't know if the value is a string, int, etc. What is the initial
query used for?

In regards to inserts, update and deletes, query parms can be used too but
probably the most value is in plain selects.
Some non select examples are
Insert into account (acct_num, customer_id) values (?, ?)
INSERT INTO films SELECT * FROM tmp_films WHERE date_prod < ?;
UPDATE employees SET sales_count = sales_count + 1 WHERE id = (SELECT
sales_person FROM accounts WHERE name =?);
DELETE FROM films WHERE producer_id IN (SELECT id FROM producers WHERE name
= ?);

Other than the "Insert into account" statement above, the other 3
statements are tailored to very specific use cases and e.g. don't have
corresponding select statements...

**Comment by restsql** [comment]
The initial query is used for metadata inspection. It's the secret sauce in the recipe. The resource def is really supplementary info for serialization (and some hierarchy info). You could change it so the initial query is fed params instead of using the plain query with a limit clause.
 
But your examples of sub selects in write statements present an entirely new challenge. The framework currently constructs write statements from scratch using the metadata and the request params. The only way it would know the subselect clause is if it is parsed from the main query, separately identified in the definition or it was provided the text in the request.

The first one assumes it's the same as the read. The second could vary but it's quite limiting. The third is most flexible but is a security hole.

Maybe we should think about the subselect as a resource.

**Comment by BernhardLenz** [comment]
Thanks. The _params# feature is definitely a good thing to have. Will take
a shot at it.

The more I'm thinking about it... The sub-queries fall under a category
"complex queries" where a crud approach doesn't make sense. Perhaps a
different resource type is needed where just the given query is executed
(including _param replacement) and where the framework doesn't
automatically try to create insert, update or delete statements... What do
you think, does this make sense?

**Comment by restsql** [comment]
Agree it's with exploring a new type of resource. 

I think the name is challenging. Getting the name right will help crisp up the concept. If it only supports direct execution vs crud operations, is it more complex? Maybe in one dimension. Maybe it could be called an operation? But that might confuse with the web service operation concept. Maybe it's a statement or an exec query? In a web service sense they are both resources - everything is. Maybe one is fixed and one is flexible? How about static and dynamic? Open or closed?

It should definitely be protectable like current resources.

**Comment by BernhardLenz** [comment]
Yeah the naming is tricky but important. both are definitely sql resources.

Maybe if the meta-data table tags are missing the framework could assume
that this is just a statement? In that case the naming could stay the
same...

The protection is a good thing and definitely applies to exec only
statements as well.

Not to side track...
I noticed that if a column value is null, restsql does not return the field
in JSON. This makes it a little difficult to transform the json object into
a table since the number of elements of each row varies. I got around it by
using postgres' "select coalesce(com_program.pgm_no,'') pgm_no, ..." for
each nullable column but this approach is a little cumbersome. Is there a
way to configure restsql to always return the null fields?

**Comment by restsql** [comment]
Lack of metadata would likely fail XML validation. Regardless I feel a more explicit declaration is better. Some new attribute is needed.

I dont think there is a config to control attribute display for null values. It's something that could be easily added to Config. The Json deserializer would read it and act accordingly.
